extensions/amp-next-page/0.1/next-page-service.js

```diff
@@ -269,53 +270,53 @@ export class NextPageService {
       }
 
       this.nextArticle_++;
-      this.xhr_.fetch(next.ampUrl, {ampCors: false})
-          .then(response => {
-            // Update AMP URL in case we were redirected.
-            documentRef.ampUrl = response.url;
-            const url = this.urlService_.parse(response.url);
-            userAssert(url.origin === this.origin_,
-                'ampUrl resolved to a different origin from the origin of the '
-                + 'current document');
-            return response.text();
-          })
-          .then(html => {
-            const doc =
-                this.win_.document.implementation.createHTMLDocument('');
-            doc.open();
-            doc.write(html);
-            doc.close();
-            return doc;
-          })
-          .then(doc => new Promise((resolve, reject) => {
-            if (documentRef.cancelled) {
-              // User has reached the end of the document already, don't render.
-              resolve();
-              return;
-            }
+      const fetchPromise = this.xhr_.fetch(ampUrl, {ampCors: false}).then(r => {
+        // Update AMP URL in case we were redirected.
+        documentRef.ampUrl = r.url;
+        const url = this.urlService_.parse(r.url);
+        userAssert(url.origin === this.origin_,
+            'ampUrl resolved to a different origin from the origin of the '
+            + 'current document');
+        return r.text();
+      }, e => {
+        user().error(TAG, 'failed to fetch %s', ampUrl, e);
+      });
+      // Once next page's HTML at `ampUrl` is fetched, inject it into a new doc.
+      fetchPromise.then(html => {
```

Comment:
I'm not sure this does what you want. The catch above will cause you to enter this `.then` with  `html` being `undefined`, and isn't any different than doing

```javascript
this.xhr_.fetch(...)
  .then(r => { ...}, e => {...})
  .then(html => [...});
```

Since this function is quite long already, I think this could be fixed and clarified by breaking out the attachment of the html part into a separate function. For example:

```javascript
const fetchPromise = this.xhr_.fetch(...);

fetchPromise.catch(e => {...});
fetchPromise
  .then(r => {...})
  .then(html => {
     return appendArticleFromHtml(html);
  });
```

As-is, it is pretty hard to follow the flow of `.then` and `.catch` due to how long the code is.

The new code seems like it should attach a doc with the content 'undefined', which I don't think is the right thing to do for a failure.

The previous code looks to behave correctly by skipping all the attaching steps in the in the case of the fetch failure by letting the error fall through until the failed to fetch dev error.

Reply:
Good catch, added an early return. 👍 

Yea I couldn't resist the light refactoring because of readability. Looked into a "appendArticleWithHtml" helper but it'd need ~5 params so it doesn't feel like a readability win.

Reply:
Added a suggestion on how to refactor this above. You don't really need to pass too many arguments, if you separate out concerns here.

There are also a few minor things to clean up. For example, there is no need to wrap the mutate element with a try/catch, because the catch only is only ever logging an error and recovering into a non-failed state (assuming the `unobserve `doesn't throw an error), you could just do:

``` javascript
if (documentRef.cancelled) {
  return;
}

...

return this.resources_.mutateElement(container, () => {
  try {
    ...
  } catch (e) {
    dev().error(...);
  }
});
```

Reply:
I would prefer either this is code is refactored more broadly, or the minimum change needed is made in this PR.

Reply:
Minimum change it is. :)